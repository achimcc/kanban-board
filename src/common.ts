enum TaskStatus {
  ToDo = "TO_DO",
  Doing = "DOING",
  Done = "DONE",
}

export { TaskStatus };
