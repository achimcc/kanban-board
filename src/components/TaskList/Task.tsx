import * as React from "react";
import { useDispatch } from "react-redux";

import { updateTask, removeTask } from "../../store/taskList/actionCreators";
import { TaskStatus } from "../../common";

import {
  Button,
  ListItem,
  ListItemIcon,
  ListItemText,
} from "@material-ui/core";
import { SkipPrevious } from "@material-ui/icons";
import SkipNextIcon from "@material-ui/icons/SkipNext";
import CloseIcon from "@material-ui/icons/Close";

interface Props {
  task: ITask;
}

enum Direction {
  Left,
  Right,
}

const moveTask = (task: ITask, direction: Direction) => {
  const statusArr = Object.values(TaskStatus);
  const statusIndex =
    statusArr.indexOf(task.status) + (direction === Direction.Left ? -1 : 1);
  const status =
    statusIndex > -1 && statusIndex < 3 ? statusArr[statusIndex] : task.status;
  return updateTask(task.id, status);
};

const Task = ({ task }: Props) => {
  const dispatch = useDispatch();
  return (
    <ListItem>
      <ListItemIcon>
        {task.status !== TaskStatus.ToDo && (
          <Button onClick={() => dispatch(moveTask(task, Direction.Left))}>
            <SkipPrevious />
          </Button>
        )}
      </ListItemIcon>
      <ListItemText primary={task.title} />
      <ListItemIcon>
        {task.status == TaskStatus.Done ? (
          <Button onClick={() => dispatch(removeTask(task.id))}>
            <CloseIcon />
          </Button>
        ) : (
          <Button onClick={() => dispatch(moveTask(task, Direction.Right))}>
            <SkipNextIcon />
          </Button>
        )}
      </ListItemIcon>
    </ListItem>
  );
};

export default Task;
