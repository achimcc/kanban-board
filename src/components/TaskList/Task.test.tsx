import { shallow, render, mount, ReactWrapper } from "enzyme";
import { TaskStatus } from "../../common";
import configureMockStore from "redux-mock-store";
import Task from "./Task";
import { Provider } from "react-redux";
import { Store } from "redux";
import { SkipPrevious, SkipNext, Close } from "@material-ui/icons";
import { ShallowRenderer } from "react-dom/test-utils";

const mockStore = configureMockStore([]);

describe("Task", () => {
  var store: Store;
  var componentFromTask: any;
  beforeEach(() => {
    store = mockStore({
      ui: {},
      tasks: {},
    });
    store.dispatch = jest.fn();
    componentFromTask = (task: ITask) =>
      mount(
        <Provider store={store}>
          <Task task={task} />
        </Provider>
      );
  });
  it("is truthy", () => {
    expect(Task).toBeTruthy();
  });

  it("to match snapshot", () => {
    const task: ITask = {
      id: 2,
      title: "test item",
      status: TaskStatus.Done,
    };
    const component = componentFromTask(task);
    expect(component).toMatchSnapshot();
  });

  it("can move left", () => {
    const task: ITask = {
      id: 2,
      title: "test item",
      status: TaskStatus.Done,
    };
    const component = componentFromTask(task);
    component.find(SkipPrevious).simulate("click");
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: "SET_STATUS",
      id: 2,
      status: TaskStatus.Doing,
    });
  });
  it("can be deleted", () => {
    const task: ITask = {
      id: 2,
      title: "test item",
      status: TaskStatus.Done,
    };
    const component = componentFromTask(task);
    expect(component).toMatchSnapshot();
    component.find(Close).simulate("click");
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith({
      type: "DELETE",
      id: 2,
    });
  });
});
