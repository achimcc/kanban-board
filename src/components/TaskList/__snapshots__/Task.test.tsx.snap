// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task can be deleted 1`] = `ReactWrapper {}`;

exports[`Task to match snapshot 1`] = `ReactWrapper {}`;
